File: Training/Advanced_Cluster_Usage/Exercise2/Fortran/external_libs.f90
SUBROUTINE TRED2(NM, N, A, D, E, Z)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: NM, N
  REAL(8), INTENT(INOUT) :: A(NM, N), Z(NM, N)
  REAL(8), INTENT(OUT) :: D(N), E(N)
  INTEGER :: I, J, K, L, II, JP1
  REAL(8) :: F, G, H, HH, SCALE

  ! Copy lower triangle of A to Z and initialize D
  DO I = 1, N
    DO J = I, N
      Z(J, I) = A(J, I)
    END DO
    D(I) = A(N, I)
  END DO

  ! Handle the special case of N = 1
  IF (N == 1) THEN
    E(1) = 0.0D0
    Z(1, 1) = 1.0D0
    RETURN
  END IF

  ! Tridiagonalization loop
  DO II = 2, N
    I = N + 2 - II
    L = I - 1
    H = 0.0D0
    SCALE = 0.0D0

    ! Compute scale factor
    IF (L >= 2) THEN
      DO K = 1, L
        SCALE = SCALE + ABS(D(K))
      END DO
    END IF

    IF (SCALE == 0.0D0) THEN
      E(I) = D(L)
      DO J = 1, L
        D(J) = Z(L, J)
        Z(I, J) = 0.0D0
        Z(J, I) = 0.0D0
      END DO
    ELSE
      ! Scale the vector and compute H
      DO K = 1, L
        D(K) = D(K) / SCALE
        H = H + D(K) * D(K)
      END DO
      F = D(L)
      G = -SIGN(SQRT(H), F)
      E(I) = SCALE * G
      H = H - F * G
      D(L) = F - G

      ! Form A*U
      DO J = 1, L
        E(J) = 0.0D0
      END DO
      DO J = 1, L
        F = D(J)
        Z(J, I) = F
        G = E(J) + Z(J, J) * F
        JP1 = J + 1
        IF (JP1 <= L) THEN
          DO K = JP1, L
            G = G + Z(K, J) * D(K)
            E(K) = E(K) + Z(K, J) * F
          END DO
        END IF
        E(J) = G
      END DO

      ! Form P and Q
      F = 0.0D0
      DO J = 1, L
        E(J) = E(J) / H
        F = F + E(J) * D(J)
      END DO
      HH = F / (H + H)
      DO J = 1, L
        E(J) = E(J) - HH * D(J)
      END DO

      ! Reduce A
      DO J = 1, L
        F = D(J)
        G = E(J)
        DO K = J, L
          Z(K, J) = Z(K, J) - F * E(K) - G * D(K)
        END DO
        D(J) = Z(L, J)
        Z(I, J) = 0.0D0
      END DO
    END IF
    D(I) = H
  END DO

  ! Accumulate transformation matrices
  DO I = 2, N
    L = I - 1
    Z(N, L) = Z(L, L)
    Z(L, L) = 1.0D0
    H = D(I)
    IF (H /= 0.0D0) THEN
      DO K = 1, L
        D(K) = Z(K, I) / H
      END DO
      DO J = 1, L
        G = 0.0D0
        DO K = 1, L
          G = G + Z(K, I) * Z(K, J)
        END DO
        DO K = 1, L
          Z(K, J) = Z(K, J) - G * D(K)
        END DO
      END DO
    END IF
    DO K = 1, L
      Z(K, I) = 0.0D0
    END DO
  END DO

  ! Finalize output
  DO I = 1, N
    D(I) = Z(N, I)
    Z(N, I) = 0.0D0
  END DO
  Z(N, N) = 1.0D0
  E(1) = 0.0D0
END SUBROUTINE TRED2

SUBROUTINE TQLI(NM, N, D, E, Z, IERR)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: NM, N
  REAL(8), INTENT(INOUT) :: D(N), E(N), Z(NM, N)
  INTEGER, INTENT(OUT) :: IERR
  INTEGER :: I, J, K, L, M, II, L1, L2, MML
  REAL(8) :: C, C2, C3, DL1, EL1, F, G, H, P, R, S, S2, TST1, TST2
  REAL(8), EXTERNAL :: PYTHAG

  IERR = 0
  IF (N == 1) RETURN

  ! Shift subdiagonal elements
  DO I = 2, N
    E(I-1) = E(I)
  END DO
  E(N) = 0.0D0
  F = 0.0D0
  TST1 = 0.0D0

  ! Main QL iteration
  DO L = 1, N
    J = 0
    H = ABS(D(L)) + ABS(E(L))
    IF (TST1 < H) TST1 = H
    M = L
    DO WHILE (M <= N)
      TST2 = TST1 + ABS(E(M))
      IF (TST2 == TST1) EXIT
      M = M + 1
    END DO
    IF (M == L) THEN
      D(L) = D(L) + F
      CYCLE
    END IF
    DO
      IF (J == 30) THEN
        IERR = L
        RETURN
      END IF
      J = J + 1
      L1 = L + 1
      L2 = L1 + 1
      G = D(L)
      P = (D(L1) - G) / (2.0D0 * E(L))
      R = PYTHAG(P, 1.0D0)
      D(L) = E(L) / (P + SIGN(R, P))
      D(L1) = E(L) * (P + SIGN(R, P))
      DL1 = D(L1)
      H = G - D(L)
      IF (L2 <= N) THEN
        DO I = L2, N
          D(I) = D(I) - H
        END DO
      END IF
      F = F + H

      ! QL transformation
      P = D(M)
      C = 1.0D0
      C2 = C
      EL1 = E(L1)
      S = 0.0D0
      MML = M - L
      DO II = 1, MML
        C3 = C2
        C2 = C
        S2 = S
        I = M - II
        G = C * E(I)
        H = C * P
        R = PYTHAG(P, E(I))
        E(I+1) = S * R
        S = E(I) / R
        C = P / R
        P = C * D(I) - S * G
        D(I+1) = H + S * (C * G + S * D(I))
        DO K = 1, N
          H = Z(K, I+1)
          Z(K, I+1) = S * Z(K, I) + C * H
          Z(K, I) = C * Z(K, I) - S * H
        END DO
      END DO
      P = -S * S2 * C3 * EL1 * E(L) / DL1
      E(L) = S * P
      D(L) = C * P
      TST2 = TST1 + ABS(E(L))
      IF (TST2 <= TST1) EXIT
    END DO
    D(L) = D(L) + F
  END DO

  ! Order eigenvalues and eigenvectors
  DO II = 2, N
    I = II - 1
    K = I
    P = D(I)
    DO J = II, N
      IF (D(J) < P) THEN
        K = J
        P = D(J)
      END IF
    END DO
    IF (K /= I) THEN
      D(K) = D(I)
      D(I) = P
      DO J = 1, N
        P = Z(J, I)
        Z(J, I) = Z(J, K)
        Z(J, K) = P
      END DO
    END IF
  END DO
END SUBROUTINE TQLI

REAL(8) FUNCTION PYTHAG(A, B)
  IMPLICIT NONE
  REAL(8), INTENT(IN) :: A, B
  REAL(8) :: P, R, S, T, U
  P = MAX(ABS(A), ABS(B))
  IF (P == 0.0D0) THEN
    PYTHAG = 0.0D0
    RETURN
  END IF
  R = (MIN(ABS(A), ABS(B)) / P)**2
  DO
    T = 4.0D0 + R
    IF (T == 4.0D0) EXIT
    S = R / T
    U = 1.0D0 + 2.0D0 * S
    P = U * P
    R = (S / U)**2 * R
  END DO
  PYTHAG = P
END FUNCTION PYTHAG

SUBROUTINE EIGSRT(D, V, N, NP)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: N, NP
  REAL(8), INTENT(INOUT) :: D(NP), V(NP, NP)
  INTEGER :: I, J, K
  REAL(8) :: P
  DO I = 1, N-1
    K = I
    P = D(I)
    DO J = I+1, N
      IF (D(J) >= P) THEN
        K = J
        P = D(J)
      END IF
    END DO
    IF (K /= I) THEN
      D(K) = D(I)
      D(I) = P
      DO J = 1, N
        P = V(J, I)
        V(J, I) = V(J, K)
        V(J, K) = P
      END DO
    END IF
  END DO
END SUBROUTINE EIGSRT

REAL(8) FUNCTION RAN3(IDUM)
  IMPLICIT NONE
  INTEGER, INTENT(INOUT) :: IDUM
  INTEGER, PARAMETER :: MBIG = 1000000000, MSEED = 161803398, MZ = 0
  REAL(8), PARAMETER :: FAC = 1.0D0 / MBIG
  INTEGER :: I, IFF, II, INEXT, INEXTP, K, MJ, MK
  INTEGER :: MA(55)
  SAVE IFF, INEXT, INEXTP, MA
  DATA IFF /0/

  IF (IDUM < 0 .OR. IFF == 0) THEN
    IFF = 1
    MJ = MSEED - ABS(IDUM)
    MJ = MOD(MJ, MBIG)
    MA(55) = MJ
    MK = 1
    DO I = 1, 54
      II = MOD(21 * I, 55)
      MA(II) = MK
      MK = MJ - MK
      IF (MK < MZ) MK = MK + MBIG
      MJ = MA(II)
    END DO
    DO K = 1, 4
      DO I = 1, 55
        MA(I) = MA(I) - MA(1 + MOD(I + 30, 55))
        IF (MA(I) < MZ) MA(I) = MA(I) + MBIG
      END DO
    END DO
    INEXT = 0
    INEXTP = 31
    IDUM = 1
  END IF
  INEXT = INEXT + 1
  IF (INEXT == 56) INEXT = 1
  INEXTP = INEXTP + 1
  IF (INEXTP == 56) INEXTP = 1
  MJ = MA(INEXT) - MA(INEXTP)
  IF (MJ < MZ) MJ = MJ + MBIG
  MA(INEXT) = MJ
  RAN3 = MJ * FAC
END FUNCTION RAN3

SUBROUTINE SHELLSORT(N, IND, DIM1, ARR)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: N, IND, DIM1
  INTEGER, INTENT(INOUT) :: ARR(DIM1, N)
  INTEGER :: I, J, INC, V
  INTEGER :: VARR(DIM1)

  INC = 1
  DO
    INC = 3 * INC + 1
    IF (INC > N) EXIT
  END DO
  DO
    INC = INC / 3
    DO I = INC + 1, N
      V = ARR(IND, I)
      VARR = ARR(:, I)
      J = I
      DO WHILE (J > INC .AND. ARR(IND, J - INC) > V)
        ARR(:, J) = ARR(:, J - INC)
        J = J - INC
      END DO
      ARR(:, J) = VARR
    END DO
    IF (INC <= 1) EXIT
  END DO
END SUBROUTINE SHELLSORT

SUBROUTINE PIKSRT(N, ARR)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: N
  REAL(8), INTENT(INOUT) :: ARR(N)
  INTEGER :: I, J
  REAL(8) :: A
  DO J = 2, N
    A = ARR(J)
    I = J - 1
    DO WHILE (I >= 1 .AND. ARR(I) > A)
      ARR(I + 1) = ARR(I)
      I = I - 1
    END DO
    ARR(I + 1) = A
  END DO
END SUBROUTINE PIKSRT

File: Training/Advanced_Cluster_Usage/Exercise2/C/external_libs.c
#include <math.h>
#include <stdlib.h>     
#include <stdio.h>

void tred2(int nm, int n, double *a, double *d, double *e, double *z) {
    int i, j, k, l, ii, jp1;
    double f, g, h, hh, scale;

    // Copy lower triangle of A to Z and initialize D
    for (i = 0; i < n; i++) {
        for (j = i; j < n; j++) {
            z[j * nm + i] = a[j * nm + i];
        }
        d[i] = a[(n - 1) * nm + i];
    }

    // Handle the special case of n = 1
    if (n == 1) {
        e[0] = 0.0;
        z[0] = 1.0;
        return;
    }

    // Tridiagonalization loop
    for (ii = 1; ii < n; ii++) {
        i = n - 1 - (ii - 1); // Adjusted for 0-based indexing
        l = i - 1;
        h = 0.0;
        scale = 0.0;

        // Compute scale factor
        if (l >= 0) { // Adjusted for 0-based indexing
            for (k = 0; k <= l; k++) {
                scale += fabs(d[k]);
            }
        }

        if (scale == 0.0) {
            e[i] = d[l];
            for (j = 0; j <= l; j++) {
                d[j] = z[l * nm + j];
                z[i * nm + j] = 0.0;
                z[j * nm + i] = 0.0;
            }
        } else {
            // Scale the vector and compute h
            for (k = 0; k <= l; k++) {
                d[k] /= scale;
                h += d[k] * d[k];
            }
            f = d[l];
            g = -copysign(sqrt(h), f);
            e[i] = scale * g;
            h -= f * g;
            d[l] = f - g;

            // Form A*U
            for (j = 0; j <= l; j++) {
                e[j] = 0.0;
            }
            for (j = 0; j <= l; j++) {
                f = d[j];
                z[j * nm + i] = f;
                g = e[j] + z[j * nm + j] * f;
                jp1 = j + 1;
                if (jp1 <= l) {
                    for (k = jp1; k <= l; k++) {
                        g += z[k * nm + j] * d[k];
                        e[k] += z[k * nm + j] * f;
                    }
                }
                e[j] = g;
            }

            // Form P and Q
            f = 0.0;
            for (j = 0; j <= l; j++) {
                e[j] /= h;
                f += e[j] * d[j];
            }
            hh = f / (h + h);
            for (j = 0; j <= l; j++) {
                e[j] -= hh * d[j];
            }

            // Reduce A
            for (j = 0; j <= l; j++) {
                f = d[j];
                g = e[j];
                for (k = j; k <= l; k++) {
                    z[k * nm + j] -= f * e[k] + g * d[k];
                }
                d[j] = z[l * nm + j];
                z[i * nm + j] = 0.0;
            }
        }
        d[i] = h;
    }

    // Accumulate transformation matrices
    for (i = 1; i < n; i++) {
        l = i - 1;
        z[(n - 1) * nm + l] = z[l * nm + l];
        z[l * nm + l] = 1.0;
        h = d[i];
        if (h != 0.0) {
            for (k = 0; k <= l; k++) {
                d[k] = z[k * nm + i] / h;
            }
            for (j = 0; j <= l; j++) {
                g = 0.0;
                for (k = 0; k <= l; k++) {
                    g += z[k * nm + i] * z[k * nm + j];
                }
                for (k = 0; k <= l; k++) {
                    z[k * nm + j] -= g * d[k];
                }
            }
        }
        for (k = 0; k <= l; k++) {
            z[k * nm + i] = 0.0;
        }
    }

    // Finalize output
    for (i = 0; i < n; i++) {
        d[i] = z[(n - 1) * nm + i];
        z[(n - 1) * nm + i] = 0.0;
    }
    z[(n - 1) * nm + (n - 1)] = 1.0;
    e[0] = 0.0;
}

double pythag(double a, double b); // Forward declaration

void tqli(int nm, int n, double *d, double *e, double *z, int *ierr) {
    int i, j, k, l, m, ii, l1, l2, mml;
    double c, c2, c3, dl1, el1, f, g, h, p, r, s, s2, tst1, tst2;

    *ierr = 0;
    if (n == 1) return;

    // Shift subdiagonal elements
    for (i = 1; i < n; i++) {
        e[i - 1] = e[i];
    }
    e[n - 1] = 0.0;
    f = 0.0;
    tst1 = 0.0;

    // Main QL iteration
    for (l = 0; l < n; l++) {
        j = 0;
        h = fabs(d[l]) + fabs(e[l]);
        if (tst1 < h) tst1 = h;
        m = l;
        while (m < n) {
            tst2 = tst1 + fabs(e[m]);
            if (tst2 == tst1) break;
            m++;
        }
        if (m == l) {
            d[l] += f;
            continue;
        }
        do {
            if (j == 30) {
                *ierr = l + 1; // Adjusted for 1-based error reporting
                return;
            }
            j++;
            l1 = l + 1;
            l2 = l1 + 1;
            g = d[l];
            p = (d[l1] - g) / (2.0 * e[l]);
            r = pythag(p, 1.0);
            d[l] = e[l] / (p + copysign(r, p));
            d[l1] = e[l] * (p + copysign(r, p));
            dl1 = d[l1];
            h = g - d[l];
            if (l2 < n) {
                for (i = l2; i < n; i++) {
                    d[i] -= h;
                }
            }
            f += h;

            // QL transformation
            p = d[m];
            c = 1.0;
            c2 = c;
            el1 = e[l1];
            s = 0.0;
            mml = m - l;
            for (ii = 0; ii < mml; ii++) {
                c3 = c2;
                c2 = c;
                s2 = s;
                i = m - 1 - ii;
                g = c * e[i];
                h = c * p;
                r = pythag(p, e[i]);
                e[i + 1] = s * r;
                s = e[i] / r;
                c = p / r;
                p = c * d[i] - s * g;
                d[i + 1] = h + s * (c * g + s * d[i]);
                for (k = 0; k < n; k++) {
                    h = z[(i + 1) * nm + k];
                    z[(i + 1) * nm + k] = s * z[i * nm + k] + c * h;
                    z[i * nm + k] = c * z[i * nm + k] - s * h;
                }
            }
            p = -s * s2 * c3 * el1 * e[l] / dl1;
            e[l] = s * p;
            d[l] = c * p;
            tst2 = tst1 + fabs(e[l]);
            if (tst2 <= tst1) break;
        } while (1);
        d[l] += f;
    }

    // Order eigenvalues and eigenvectors
    for (ii = 1; ii < n; ii++) {
        i = ii - 1;
        k = i;
        p = d[i];
        for (j = ii; j < n; j++) {
            if (d[j] < p) {
                k = j;
                p = d[j];
            }
        }
        if (k != i) {
            d[k] = d[i];
            d[i] = p;
            for (j = 0; j < n; j++) {
                p = z[i * nm + j];
                z[i * nm + j] = z[k * nm + j];
                z[k * nm + j] = p;
            }
        }
    }
}

double pythag(double a, double b) {
    double p, r, s, t, u;
    p = fmax(fabs(a), fabs(b));
    if (p == 0.0) {
        return 0.0;
    }
    r = pow(fmin(fabs(a), fabs(b)) / p, 2);
    while (1) {
        t = 4.0 + r;
        if (t == 4.0) break;
        s = r / t;
        u = 1.0 + 2.0 * s;
        p *= u;
        r = pow(s / u, 2) * r;
    }
    return p;
}

void eigsrt(double *d, double *v, int n, int np) {
    int i, j, k;
    double p;
    for (i = 0; i < n - 1; i++) {
        k = i;
        p = d[i];
        for (j = i + 1; j < n; j++) {
            if (d[j] >= p) {
                k = j;
                p = d[j];
            }
        }
        if (k != i) {
            d[k] = d[i];
            d[i] = p;
            for (j = 0; j < n; j++) {
                p = v[i * np + j];
                v[i * np + j] = v[k * np + j];
                v[k * np + j] = p;
            }
        }
    }
}

double ran3(int *idum) {
    const int MBIG = 1000000000;
    const int MSEED = 161803398;
    const int MZ = 0;
    const double FAC = 1.0 / MBIG;
    static int iff = 0;
    static int inext, inextp;
    static int ma[55];
    int i, ii, k, mj, mk;

    if (*idum < 0 || iff == 0) {
        iff = 1;
        mj = MSEED - abs(*idum);
        mj %= MBIG;
        ma[54] = mj;
        mk = 1;
        for (i = 0; i < 54; i++) {
            ii = (21 * (i + 1)) % 55;
            ma[ii - 1] = mk;
            mk = mj - mk;
            if (mk < MZ) mk += MBIG;
            mj = ma[ii - 1];
        }
        for (k = 0; k < 4; k++) {
            for (i = 0; i < 55; i++) {
                ma[i] -= ma[(i + 31) % 55];
                if (ma[i] < MZ) ma[i] += MBIG;
            }
        }
        inext = 0;
        inextp = 31;
        *idum = 1;
    }
    inext = (inext + 1) % 55;
    inextp = (inextp + 1) % 55;
    mj = ma[inext] - ma[inextp];
    if (mj < MZ) mj += MBIG;
    ma[inext] = mj;
    return mj * FAC;
}

void shellsort(int n, int ind, int dim1, int *arr) {
    int i, j, inc;
    double v;
    double *varr = malloc(dim1 * sizeof(double)); // Temporary array

    inc = 1;
    while (inc <= n / 3) {
        inc = 3 * inc + 1;
    }
    while (inc > 0) {
        for (i = inc; i < n; i++) {
            v = arr[ind * dim1 + i];
            for (j = 0; j < dim1; j++) {
                varr[j] = arr[j * dim1 + i];
            }
            j = i;
            while (j >= inc && arr[ind * dim1 + (j - inc)] > v) {
                for (int k = 0; k < dim1; k++) {
                    arr[k * dim1 + j] = arr[k * dim1 + (j - inc)];
                }
                j -= inc;
            }
            for (int k = 0; k < dim1; k++) {
                arr[k * dim1 + j] = varr[k];
            }
        }
        inc /= 3;
    }
    free(varr);
}

void piksrt(int n, double *arr) {
    int i, j;
    double a;
    for (j = 1; j < n; j++) {
        a = arr[j];
        i = j - 1;
        while (i >= 0 && arr[i] > a) {
            arr[i + 1] = arr[i];
            i--;
        }
        arr[i + 1] = a;
    }
}


